File: CalcByBakosta/app/src/main/java/com/example/calcbybakosta/Calculator.java
package com.example.calcbybakosta;

import java.util.Stack;

public class Calculator {


    public enum Operator{ADD, SUBTRACT, MULTIPLY, DIVIDE, BLANK, DOT, DBLE, SQRT, FACT, PERCENT, SIN, COS,TAN,LN,LOG,E,XN,SQRTN}

    public static void main(String[] args){
        String expression = "2-6-7*8/2+5";
        Calculator calc = new Calculator();
        System.out.println(calc.compute(expression));
    }

    public double compute(String sequence){
        Stack<Double> numberStack = new Stack<Double>();
        Stack<Operator> operatorStack = new Stack<Operator>();
        for(int i = 0; i < sequence.length(); i++){
            try{
                int number = parseNumber(sequence, i);
                numberStack.push((double)number);

                i += Integer.toString(number).length();
                if(i >= sequence.length()){
                    break;
                }

                Operator op = parseOperator(sequence, i);
                collapseTop(numberStack, operatorStack, op);
                operatorStack.push(op);
            } catch(NumberFormatException ex){
                return Integer.MIN_VALUE;
            }
        }
        collapseTop(numberStack, operatorStack, Operator.BLANK);
        if(numberStack.size() == 1 && operatorStack.size() == 0){
            return numberStack.pop();
        }
        return 0;
    }
//operatorsStack peek futureTop
    private void collapseTop(Stack<Double> numberStack, Stack<Operator> operatorStack, Operator futureTop){
        while(numberStack.size() >= 2 && operatorStack.size() >= 1){
            if(priorityOfOperator(futureTop) <= priorityOfOperator(operatorStack.peek())){

                double second = numberStack.pop();
                double first = numberStack.pop();

                Operator op = operatorStack.pop();

                double result = applyOp(first, op, second);

                numberStack.push(result);
            } else{
                break;
            }
        }
    }
    private double setMaxSize(double answer){
        double divSum = Double.parseDouble(String.format("%.3f",answer));
        return divSum;

    }
    private double applyOp(double left, Operator op, double right){
        int left1 = (int)left;
        int right1 = (int)right;//secret 2.06!

        String dotcase = left1 + "." + right1;
        Double res = Double.parseDouble(dotcase);
        switch (op){
//            case ADD: return left + right;
//            case SUBTRACT: return left - right;
//            case MULTIPLY: return left * right;
//            case DIVIDE: return left / right;
//            case DBLE: return left * left;
//            case DOT: return res;
//            case SQRT: return Math.sqrt(left);
//            case FACT: return fact(left);
            case ADD: return setMaxSize(left + right);
            case SUBTRACT: return setMaxSize(left - right);
            case MULTIPLY: return setMaxSize(left * right);
            case DIVIDE: return setMaxSize(left / right);
            case DBLE: return setMaxSize(left * left);
            case DOT: return res;
            case SQRT: return setMaxSize(Math.sqrt(left));
            case FACT: return setMaxSize(fact(left));


            case PERCENT: return right*left/100;
            case SIN:return Math.sin(right);
            case COS:return Math.cos(right);
            case TAN:return Math.tan(right);
            case LN:return Math.log(right);
            case LOG:return Math.log10(right);
            case E:return 2.7182;
            case XN:return Math.pow(left,right);
            case SQRTN:return sqrtN(left,right);
            default: return right;
        }

    }

    private int priorityOfOperator(Operator op){
        switch (op){
            case ADD: return 1;
            case SUBTRACT: return 1;
            case MULTIPLY: return 2;
            case DIVIDE: return 2;
            case DBLE: return 3;
            case FACT: return 3;
            case SQRT: return 3;
            case DOT: return 4;
            case BLANK: return 0;

            case PERCENT:return 3;
            case SIN:return 3;
            case COS:return 3;
            case TAN:return 3;
            case LN:return 3;
            case LOG:return 3;
            case E:return 5;
            case XN:return 2;
            case SQRTN:return 3;

        }
        return 0;
    }

    private int parseNumber(String sequence, int offset){
        StringBuilder sb = new StringBuilder();
        while(offset < sequence.length() && Character.isDigit(sequence.charAt(offset))){
            sb.append(sequence.charAt(offset));
            offset++;
        }
        return Integer.parseInt(sb.toString());
    }

    private Operator parseOperator(String sequence, int offset){
        if(offset < sequence.length()){
            char op = sequence.charAt(offset);
            switch (op){
                case '+': return Operator.ADD;
                case '-': return Operator.SUBTRACT;
                case '*': return Operator.MULTIPLY;
                case '/': return Operator.DIVIDE;
                case '.': return Operator.DOT;
                case 'x': return Operator.DBLE;
                case '√': return Operator.SQRT;
                case '!': return Operator.FACT;



                case '%': return Operator.PERCENT;
                case 's': return Operator.SIN;
                case 'c': return Operator.COS;
                case 't': return Operator.TAN;
                case 'l': return Operator.LN;
                case 'g': return Operator.LOG;
                case 'e': return Operator.E;
                case '^': return Operator.XN;
                case 'n': return Operator.SQRTN;

            }
        }
        return Operator.BLANK;
    }
    private double sqrtN(double left, double right){
        if(right==0){
            return 1;
        }else
            return Math.pow(left, 1.0/right);
    }
    private long fact(double left){
        int n = (int)left;
        long fact=1;
        if(n==0)return 1;
        else{
            for(int i=1;i<=n;i++)
            {
                fact=fact*i;
            }
            return fact;
        }

    }
}

File: CalcInKotlin/CalcByBakosta/app/src/main/java/com/example/calcbybakosta/Calculator.kt
package com.example.calcbybakosta

import java.util.*

class Calculator {
    enum class Operator {
        ADD, SUBTRACT, MULTIPLY, DIVIDE, BLANK, DOT, DBLE, SQRT, FACT, PERCENT, SIN, COS, TAN, LN, LOG, E, XN, SQRTN
    }

    fun compute(sequence: String): Double {
        val numberStack = Stack<Double>()
        val operatorStack = Stack<Operator>()
        var i = 0
        while (i < sequence.length) {
            try {
                val number = parseNumber(sequence, i)
                numberStack.push(number.toDouble())
                i += Integer.toString(number).length
                if (i >= sequence.length) {
                    break
                }
                val op = parseOperator(sequence, i)
                collapseTop(numberStack, operatorStack, op)
                operatorStack.push(op)
            } catch (ex: NumberFormatException) {
                return Int.MIN_VALUE.toDouble()
            }
            i++
        }
        collapseTop(numberStack, operatorStack, Operator.BLANK)
        return if (numberStack.size == 1 && operatorStack.size == 0) {
            numberStack.pop()
        } else
            return 0.0
    }

    //operatorsStack peek futureTop
    private fun collapseTop(numberStack: Stack<Double>, operatorStack: Stack<Operator>, futureTop: Operator) {
        while (numberStack.size >= 2 && operatorStack.size >= 1) {
            if (priorityOfOperator(futureTop) <= priorityOfOperator(operatorStack.peek())) {
                val second = numberStack.pop()
                val first = numberStack.pop()
                val op = operatorStack.pop()
                val result = applyOp(first, op, second)
                numberStack.push(result)
            } else {
                break
            }
        }
    }

    private fun setMaxSize(answer: Double): Double {
        return String.format("%.3f", answer).toDouble()
    }

    private fun applyOp(left: Double, op: Operator, right: Double): Double {
        val left1 = left.toInt()
        val right1 = right.toInt() //secret 2.06!
        val dotcase = "$left1.$right1"
        val res = dotcase.toDouble()
        return when (op) {
            Operator.ADD -> setMaxSize(left + right)
            Operator.SUBTRACT -> setMaxSize(left - right)
            Operator.MULTIPLY -> setMaxSize(left * right)
            Operator.DIVIDE -> setMaxSize(left / right)
            Operator.DBLE -> setMaxSize(left * left)
            Operator.DOT -> res
            Operator.SQRT -> setMaxSize(Math.sqrt(left))
            Operator.FACT -> setMaxSize(fact(left).toDouble())
            Operator.PERCENT -> right * left / 100
            Operator.SIN -> Math.sin(right)
            Operator.COS -> Math.cos(right)
            Operator.TAN -> Math.tan(right)
            Operator.LN -> Math.log(right)
            Operator.LOG -> Math.log10(right)
            Operator.E -> 2.7182
            Operator.XN -> Math.pow(left, right)
            Operator.SQRTN -> sqrtN(left, right)
            else -> right
        }
    }

    private fun priorityOfOperator(op: Operator): Int {
        return when (op) {
            Operator.ADD -> 1
            Operator.SUBTRACT -> 1
            Operator.MULTIPLY -> 2
            Operator.DIVIDE -> 2
            Operator.DBLE -> 3
            Operator.FACT -> 3
            Operator.SQRT -> 3
            Operator.DOT -> 4
            Operator.BLANK -> 0
            Operator.PERCENT -> 3
            Operator.SIN -> 3
            Operator.COS -> 3
            Operator.TAN -> 3
            Operator.LN -> 3
            Operator.LOG -> 3
            Operator.E -> 5
            Operator.XN -> 2
            Operator.SQRTN -> 3
        }
        return 0
    }

    private fun parseNumber(sequence: String, offset: Int): Int {
        var offset = offset
        val sb = StringBuilder()
        while (offset < sequence.length && Character.isDigit(sequence[offset])) {
            sb.append(sequence[offset])
            offset++
        }
        return sb.toString().toInt()
    }

    private fun parseOperator(sequence: String, offset: Int): Operator {
        if (offset < sequence.length) {
            val op = sequence[offset]
            when (op) {
                '+' -> return Operator.ADD
                '-' -> return Operator.SUBTRACT
                '*' -> return Operator.MULTIPLY
                '/' -> return Operator.DIVIDE
                '.' -> return Operator.DOT
                'x' -> return Operator.DBLE
                '√' -> return Operator.SQRT
                '!' -> return Operator.FACT
                '%' -> return Operator.PERCENT
                's' -> return Operator.SIN
                'c' -> return Operator.COS
                't' -> return Operator.TAN
                'l' -> return Operator.LN
                'g' -> return Operator.LOG
                'e' -> return Operator.E
                '^' -> return Operator.XN
                'n' -> return Operator.SQRTN
            }
        }
        return Operator.BLANK
    }

    private fun sqrtN(left: Double, right: Double): Double {
        return if (right == 0.0) {
            return 1.0
        } else Math.pow(left, 1.0 / right)
    }

    private fun fact(left: Double): Long {
        val n = left.toInt()
        var fact: Long = 1
        return if (n == 0) 1 else {
            for (i in 1..n) {
                fact = fact * i
            }
            fact
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val expression = "2-6-7*8/2+5"
            val calc = Calculator()
            println(calc.compute(expression))
        }
    }
}